### ISSUE TITLE
ASG loss does not converge

### ISSUE THREAD

**Issue by Chuhanxx** [facebookresearch/wav2letter#505]
I used to train my custom model with CTC loss: 
` loss = ctc_loss(input, target, input_lengths, target_lengths)`
and the model converges and PER is low. 

But when I switch to ASG loss using :
`loss = asg.forward(input, target, target_size)`
I noticed the loss does not decrease anymore and the PER is always close to 1.

As I didn't change anything in the model apart from the criterion,  the major difference between these two losses is that ASG does not specify the lengths of input to exclude the padded frame. I am wondering do I have to specify it somewhere? 

And I suppose the input of ASG should be the output directly from the classifier without softmax, is it right?

**Comment by tlikhomanenko** [comment]
@Chuhanxx,

Do you use C++ or python bindings?

**Comment by Chuhanxx** [comment]
> @Chuhanxx,
> 
> Do you use C++ or python bindings?
> Could you add a piece of code for model creation and how you use exactly the criterion: are you updating gradients for criterion? (because we have transitions matrix, which we learn in the ASG).

Hi, I think I am using python bindings 

My codes are:
```
from wav2letter.criterion import ASGLoss

asg_loss = ASGLoss(args.classes,scale_mode=CriterionScaleMode.TARGET_SZ_SQRT).to(device)
.......
from ... import network
model =network(args)
model = torch.nn.DataParallel(model).to(device)
out = model(input)
loss = asg_loss.forward(out, labels.long(),label_lens).mean()
optimizer.zero_grad()
loss.backward()
torch.nn.utils.clip_grad_norm_(net.parameters(), args.max_norm)
optimizer.step()
```

**Comment by xuqiantong** [comment]
@Chuhanxx  You should use `--linseg` for at least 1 epoch at the very beginning.

**Comment by tlikhomanenko** [comment]
Closing due to inactivity. Feel free to reopen!

**Comment by GCQiu** [comment]
hi ,i met the same problem,the asg loss does not decrease or converge, and the model outputs wrong results.have u solved this ？my model is wav2vec2.
